framework/Optimizers/stepManipulators/ConjugateGradient.py

```diff
@@ -0,0 +1,436 @@
+    if stepDistance > minStepSize:
+      # cut step again
+      stepSize = 0.5 * stepDistance # TODO user option?
+      for v, var in enumerate(stepVector):
+        proposed[var] = previous[var] + stepSize * stepDirection[v]
+      print(' ... cutting step ...') # norm step to {}, new norm opt {}'.format(stepSize, proposed))
+      return proposed, stepSize, fixInfo
+    else:
+      ### rotate vector and restore full step size
+      stepSize = fixInfo['originalStepSize']
+      # store original direction
+      if 'originalDirection' not in fixInfo:
+        fixInfo['originalDirection'] = np.atleast_1d(stepDirection)
+      # if this isn't the first time, check if there's angle left to rotate through; reset if not
+      if 'perpDir' in fixInfo:
+        ang = mathUtils.angleBetweenVectors(stepDirection, fixInfo['perpDir'])
+        print(' ... trying angle:', ang)
+        if ang < self._minRotationAngle:
+          del fixInfo['perpDir']
+
+      if 'perpDir' not in fixInfo:
+        # find perpendicular vector
+        perp = randomUtils.randomPerpendicularVector(fixInfo['originalDirection'])
+        # NOTE we could return to point format, but no reason to
+        # normalize perpendicular to versor and resize
+        rotations = fixInfo.get('numRotations', 0)
+        if rotations > self._numRandomPerp:
+          raise NoConstraintResolutionFound
+        _, perpDir, _ = mathUtils.calculateMagnitudeAndVersor(perp)
+        fixInfo['perpDir'] = perpDir
+        fixInfo['numRotations'] = rotations + 1
+      # END fixing perpendicular direction
+      # rotate vector halfway towards perpendicular
+      perpDir = fixInfo['perpDir']
+
+      # rotate towards selected perpendicular
+      splitVector = {} # vector that evenly divides stepDirection and perp
+      for v, var in enumerate(self._optVars):
+        splitVector[var] = stepDirection[v] + perpDir[v]
+        #splitVector[var] = - stepDirection[v] + perpDir[v]
+      _, splitDir, _ = mathUtils.calculateMagnitudeAndVersor(list(splitVector.values()))
+      for v, var in enumerate(self._optVars):
+        proposed[var] = previous[var] + stepSize * splitDir[v]
+      print(' ... rotating step ...') #ed norm direction to {}, new norm opt {}'.format(splitDir, proposed))
+    return proposed, stepSize, fixInfo
+
+  def needDenormalized(self):
+    """
+      Determines if this algorithm needs denormalized input spaces
+      @ In, None
+      @ Out, needDenormalized, bool, True if normalizing should NOT be performed
+    """
+    return True
+
+  def updateSolutionExport(self, stepHistory):
+    """
+      Prints information to the solution export.
+      @ In, stepHistory, list, (magnitude, versor, info) for each step entry
+      @ Out, info, dict, realization of data to go in the solutionExport object
+    """
+    lastStepInfo = stepHistory[-1]['info']
+    if lastStepInfo is not None:
+      task = lastStepInfo['prev task']
+      info = {'CG_task': task,
+            }
```

Comment:
reformat ?

Reply:
this format is very easy to extend; seems more like preference than standard maybe?